### PR TITLE
[Calyx] Add calyx primitive operation using hw.module.extern

### PR DESCRIPTION
**Pull request by andrewb1999**
Attempt to solve #2923, using the hw.module.extern operation to represent primitives as suggested. There are two known issues with this implementation currently:

1. Requires a modification to the HWModuleExtern Op to allow it to be used within a Calyx Program Op. Currently this is done by just removing the requirement that HWModuleExtern Ops must have a ModuleOp as a parent. Let me know if there is a better way to handle this issue.
2. HWModuleExtern does not natively support attributes per port in the same way Calyx ComponentOp does. These attributes will be required in the future to fully support external primitives. Please let me know what the correct course of action is here. The two paths I have considered are modifying HWModuleExtern to support these per port attributes or developing some encoding for these per port attributes in the Op attributes.

### CODE REVIEW

**Comment by andrewb1999** [comment]
CC: @rachitnigam

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:1355]: Should be a HWModuleExternOp?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:1378]: Why is there a default case? Why not use all the hw::PortDirections?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:1392]: ++i

https://llvm.org/docs/CodingStandards.html#prefer-preincrement

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:1469]: I *think* `value.isa<A, B, C, ...>` will work here.

File: include/circt/Dialect/HW/HWStructure.td
Comment (cgyurgyik) [inline include/circt/Dialect/HW/HWStructure.td:19]: What if we don't include the external HW operation directly in the calyx.program? Does that have different repercussions? At least in the short term, that would have the benefit of not affecting the HW dialect.

File: test/Dialect/Calyx/emit-custom-primitives.mlir
Comment (cgyurgyik) [inline test/Dialect/Calyx/emit-custom-primitives.mlir:1]: Can you add a test for this in https://github.com/llvm/circt/blob/main/test/Dialect/Calyx/round-trip.mlir as well please?

Comment (cgyurgyik) [inline test/Dialect/Calyx/emit-custom-primitives.mlir:1]: Similarly, it would be nice to verify some of the errors you've spelled out in https://github.com/llvm/circt/blob/main/test/Dialect/Calyx/errors.mlir

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:672]: Perhaps llvm_unreachable here instead of doing nothing?

https://llvm.org/docs/CodingStandards.html#assert-liberally

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:639]: type here please.

File: include/circt/Dialect/Calyx/Calyx.td
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/Calyx.td:39]: While not entirely necessary, I'd add a comment on *why* we are dependent on this.

File: include/circt/Dialect/Calyx/CalyxStructure.td
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxStructure.td:248]: At least for now, this is a HW external primitive, and not a component, right? Maybe spell that out in this example as well?

**Review by cgyurgyik — COMMENTED** [review]
Really nice work! A few comments to begin.

File: include/circt/Dialect/HW/HWStructure.td
Comment (andrewb1999) [reply include/circt/Dialect/HW/HWStructure.td:19]: I'm not sure how this would affect the CalyxEmitter. Currently the CalyxEmitter only considers operations within a calyx.program from what I understand.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp:1378]: calyx::Direction does not support InOut ports so I'm not sure how to handle the hw::PortDirection::INOUT case. Any suggestions here?

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp:1378]: Probably some assertion/compiler failure then. I tend to avoid default for enumeration switches because it may lead to buggy behavior when a new enumeration is added.

**Comment by mortbopet** [comment]
Thank you for taking a look at this! I'll give it a more thorough review later - here's my thoughts on the top-level questions

> Requires a modification to the HWModuleExtern Op to allow it to be used within a Calyx Program Op. Currently this is done by just removing the requirement that HWModuleExtern Ops must have a ModuleOp as a parent. Let me know if there is a better way to handle this issue.

I'm assuming this is needed due to `calyx.program` being `IsolatedFromAbove` (that is, a `HWModuleExternOp` in `module` scope cannot be referenced inside the `calyx.program` due to the isolation).  
Personally, i'd rather remove the `calyx.program` operation entirely, for two reasons:
1. Would allow `calyx` to compose better with other dialects - this is MLIR and that behavior should be encouraged.
2. The only thing `calyx.program` gives us is an entry point. I don't think this is necessary - if anything, that information can be derived based on the instance hierarchy (like any other hardware dialect). 

@rachitnigam, @cgyurgyik  am i missing something obvious here in the Calyx design, or is this an artifact from modeling the Calyx (ML)IR based on the native IR?

> HWModuleExtern does not natively support attributes per port in the same way Calyx ComponentOp does. These attributes will be required in the future to fully support external primitives. Please let me know what the correct course of action is here. The two paths I have considered are modifying HWModuleExtern to support these per port attributes or developing some encoding for these per port attributes in the Op attributes.

Hmm, [this](https://github.com/llvm/circt/blob/main/test/Dialect/HW/basic.mlir#L171) to me looks like per-port-attributes are supported - would we need something drastically different from that style of support?

**Comment by cgyurgyik** [comment]
> @rachitnigam, @cgyurgyik am i missing something obvious here in the Calyx design, or is this an artifact from modeling the Calyx (ML)IR based on the native IR?

Frankly, I don't recall why I added `calyx.program`. I imagine just nesting it in a ModuleOp scope would be OK. It would require refactoring quite some code, to include the CalyxEmitter as pointed out by Andrew.

**Comment by rachitnigam** [comment]
I _think_ it was mostly added to have an obvious place to get the entrypoint from. I agree that we should remove it and just encode the entrypoint information somewhere else.

**Comment by andrewb1999** [comment]
> Hmm, [this](https://github.com/llvm/circt/blob/main/test/Dialect/HW/basic.mlir#L171) to me looks like per-port-attributes are supported - would we need something drastically different from that style of support?

Looks like you're right. Thanks for pointing that out, I obviously missed it. I will work on adding attribute emission while addressing the rest of the comments.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (rachitnigam) [inline lib/Dialect/Calyx/CalyxOps.cpp:18]: This seems to include `HWAttributes` twice

File: include/circt/Dialect/HW/HWStructure.td
Comment (mikeurbach) [reply include/circt/Dialect/HW/HWStructure.td:19]: I think the CalyxEmitter can just as easily consider for builtin::ModuleOp as it can look for calyx::ProgramOp.

**Review by mikeurbach — COMMENTED** [review]
First off, thanks for pushing on this! I think this will be a great addition. A couple thoughts:

I don't think we should make any changes to the HW dialect in order to support this. If anything, Calyx should be changing to compose with the HW dialect more naturally, not the other way around. My rationale is: the HW dialect is used in production by at least two companies, making it more battle tested. What we are doing here with the Calyx dialect is more experimental, so we shouldn't let the experiments wag the dog.

As a starting point, working with HWModuleExtern is great, and already useful. But, as I mentioned on the issue, it would be great if "primitive definitions" didn't need to be written in Verilog (as is the case with a HWModuleExtern). Landing this PR as a starting point is already a step in the right direction, but I'd be curious what it would look like to replace HWModuleExtern with HWModuleLike (or even ModuleLike). I.e., anything that implements the interface could be plugged in to Calyx as a primitive. That would allow for really interesting extensions, where CIRCT-generated IR (coming from Chisel, PyCDE, or HLS) could be used for Calyx primitives, and no one would ever have to write Verilog. That is what I was hoping to work towards in the long run.

File: include/circt/Dialect/Calyx/CalyxStructure.td
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxStructure.td:17]: `MSFTModuleExternOp` typo

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1334]: Looks like this function is virtually the same as `InstanceOp::verifySymbolUses` - please merge them, if possible, to make things more DRY.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1402]: I don't want to block progress too much here, but as above, it seems like `parseParameterList`, `checkParameterInContext`, ... are similar (identical?) to what's in the HW dialect. Shared logic could be moved to something like `Support/printerparsers.h` to share parameter-printer/parsing logic - @mikeurbach thoughts?

**Review by mortbopet — CHANGES_REQUESTED** [review]

File: test/Dialect/Calyx/emit-custom-primitives.mlir
Comment (mortbopet) [inline test/Dialect/Calyx/emit-custom-primitives.mlir:39]: For the calyx port identifiers, i'd suggest using [`UnitAttr`](https://mlir.llvm.org/docs/Dialects/Builtin/#unitattr).

Comment (andrewb1999) [reply test/Dialect/Calyx/emit-custom-primitives.mlir:39]: I don't disagree, but that is not how components or the native compiler work currently. I'm trying to be consistent with the rest of the Calyx dialect for now. We can reconsider this change later.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp]: Although these functions are quite similar, the error messages have to be different depending on whether it is a primitive or a component. I can't come up with a good way to merge these functions with less code than is currently required.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: I'm not sure of the best way to do this. I would be totally okay with moving this to some shared location, I was just copying how the MSFT dialect handles ParamDeclAttr parsing. It would make a lot of sense to have both MSFT and Calyx reuse code from the HW dialect though. If I were to put this code in op interfaces instead, what would that look like? Are there any good examples of doing something similar in other dialects?

Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: I would also prefer being able to get this work merged and then fixing the duplicate code in a follow up PR, as that change will also be making changes to the MSFT and HW dialects.

File: test/Dialect/Calyx/emit-custom-primitives.mlir
Comment (andrewb1999) [reply test/Dialect/Calyx/emit-custom-primitives.mlir:39]: Nevermind, I'm wrong here, I was looking at bad examples of the Calyx dialect. I will change this to use UnitAttr.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mikeurbach) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: I think you'd start by adding the shared methods to HWInstanceLike: https://github.com/llvm/circt/blob/d8a8bf1296dbc5e740ba2a5e8e265038c49b195d/include/circt/Dialect/HW/HWOpInterfaces.td#L122

Then, you'd make sure the new PrimitiveOp uses the trait `DeclareOpInterfaceMethods<HWInstanceLike>` in the ODS file.

You'd of course need to implement them somewhere. The best bet is probably to provide a default implementation in the HWInstanceLike interface. That way they could be shared between Calyx Primitive and HW Instance (and probably MSFT Instance).

I don't really have a preference on whether this gets refactored before or after the current PR, but that change is independent. Usually (in the LLVM community), when there's a new feature and a refactor is coming to enable that, we send the refactor first, so the new feature gets to use the newly refactored thing, and both PRs are scoped to just what they are changing.

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (mikeurbach) [inline lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: This shouldn't really be necessary. Every Attribute should have a `getDialect` method, so you should be able to say `isa<CalyxDialect>(attr.getDialect())`, or something like that. That would probably be preferable to this string munging.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mikeurbach) [inline lib/Dialect/Calyx/CalyxOps.cpp:1279]: Would returning `hw::HWModuleLike` work? I think that can be generalized pretty easily as a follow up, just curious if it could work off the bat. This way, the emitter would depend on an interface, not an implementation. The interface is relatively new, so if this is a pain today, no worries.

Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp:1279]: HWModuleLike doesn't seem to have any way to access it's ports, so I don't think it's usable for now.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: I'm not sure these parser and printers make sense in the HWInstanceLike interface itself, as they aren't tied to a specific HW instance. Looking at the [ViewLikeInterface](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Interfaces/ViewLikeInterface.h#L89) these custom parsers and printers for attributes are included in the .h file, but not declared in the .td.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mikeurbach) [reply lib/Dialect/Calyx/CalyxOps.cpp:1279]: Sounds good. I guess some of those ended up in HWMutableModuleLike, but we should just make sure HWModuleLike has what it needs here before doing this generalization. Thanks for taking a look.

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (mikeurbach) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: Hmm well I guess it depends on the context. Attributes don't inherently have names. But, there are NamedAttributes, which are exposed through the DictionaryAttr class. And every Operation has a DictionaryAttr to hold its top level attributes. So in the context of an Operation, if you can get to a NamedAttribute, you can use getName on it.

Sorry if that's not very specific, I can try to point out how you'd use that here if it's still not clear.

Comment (andrewb1999) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: Yeah these are NamedAttributes, but when I use getName it returns `calyx.clk` instead of just `clk`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mikeurbach) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: I'm pretty sure this code is duplicated for the use case of parsing/printing HWInstanceLike things right (msft.instance and now calyx.primitive)? That's why I suggested putting it in an OpInterface--the two users should already be depending on a shared interface. Morten's suggestion to put them into their own header/library also makes sense.

But, I think we've actually wanted to move the parsing/printing logic to the HWModuleLike and HWInstanceLike interfaces. Not just parameters, but the whole module/instance parsing logic. That allows us to have the same textual assembly format for the ops implementing the interface, regardless of the dialect.

This isn't a major thing, and it's a bigger ask than what you're doing here, so I agree with your original point--let's leave that for a follow up, and accept the duplication for now.

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (mikeurbach) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: I see. I guess not that helpful. I think there is a `getNameDialect` that returns `CalyxDialect`, so you could do something like `attr.getName().getValue().consume_front(a.getNameDialect()->getNamespace())`.

Comment (mikeurbach) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: Just curious, why is this part changing? Is it because we are using the HW interface for calyx.primitive? 

Stepping back, didn't these attributes have the `calyx.` prefix before? Is this just filtering out attributes that aren't in the Calyx dialect?

Comment (andrewb1999) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:67]: Component port attributes do not have the `calyx.` prefix, they just have port attributes like `clk` instead of `calyx.clk`. `hw.module.extern` (and other function interfaces) require dialect attribute names to prevent collision between attribute names when dialects are mixed, therefore the port attributes for the `hw.module.extern` have to be prefixed with `calyx.`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (rachitnigam) [reply lib/Dialect/Calyx/CalyxOps.cpp:1396]: Agreed! Seems like we can open an issue to track this duplication after this is merged and clean things up afterwards

**Comment by andrewb1999** [comment]
I think I've addressed all the comments now, let me know how it looks.

**Review by cgyurgyik — APPROVED** [review]
Nice work. LGTM with a few comments left to address, and opening an issue to address the code duplication mentioned in the comment thread above. Given the size of this change, let's get some more eyes on it before merging.

**Review by mortbopet — APPROVED** [review]
LGTM, assuming that an issue is also opened up for moving this support onto `HWModuleLike` (+ the necessary changes to that interface, which sounded like it is mainly related to port attributes).

File: lib/Dialect/Calyx/Export/CalyxEmitter.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/Export/CalyxEmitter.cpp:108]: This is the way it was before, I just added the PrimitiveOp check. I assume this is because all primitives have a cell interface, so if there is no CellInterface we don't need to worry about imports. This is a little confusing because the PrimitiveOp is also not a core calyx primitive so it does not need to be considered for Calyx imports either.

**Comment by andrewb1999** [comment]
Opened an issue for the code duplication: #3629. @cgyurgyik who else should we have look at this besides Mike and Rachit? I would be happy to have more eyes on this before it is merged.

**Review by rachitnigam — APPROVED** [review]
LGTM! Other reviews seem to have covered code related things already!

**Review by mikeurbach — APPROVED** [review]
I think Morten's review is sufficient eyes on it to merge, but since I just took another look, I'll give my 👍 as well.

**Comment by andrewb1999** [comment]
@cgyurgyik Good to merge?

**Comment by cgyurgyik** [comment]
> @cgyurgyik Good to merge?

Yes, LGTM

**Comment by dtzSiFive** [comment]
FWIW seeing build failure on [CI](https://github.com/llvm/circt/runs/7674401379?check_suite_focus=true) and locally due to `std::optional`.  Didn't see obvious reason to not use `llvm::Optional` so [migrated to that](https://github.com/llvm/circt/commit/b2dbaeeae577aae51dc6de6b53c64e23a304d9fc) to fix.  LMK if `std::optional` is important/intentional, just wanted to fix the build :+1:.

**Pull request merged**
